src/app/shared/components/quantity-input/quantity-input.component.ts

```diff
@@ -0,0 +1,114 @@
+import { Component, Input } from "@angular/core";
+import { AbstractControl, ControlValueAccessor, NG_VALIDATORS, NG_VALUE_ACCESSOR, ValidationErrors, Validator } from "@angular/forms";
+
+@Component({
+    selector: "app-quantity-input",
+    templateUrl: "./quantity-input.component.html",
+    styleUrls: ["./quantity-input.component.scss"],
+    providers: [
+        {
+            provide: NG_VALUE_ACCESSOR,
+            multi: true,
+            useExisting: QuantityInputComponent,
+        },
+        {
+            provide: NG_VALIDATORS,
+            multi: true,
+            useExisting: QuantityInputComponent,
+        }
+    ]
+})
+export class QuantityInputComponent implements ControlValueAccessor, Validator {
+    @Input() label: string = "";
+    @Input() maxValue?: number;
+    @Input() minValue?: number;
+    readonly defaultValue = 0;
+    _quantity: number = this.defaultValue;
+
+    private onChange = (_: any) => {};
+    private onTouched = () => {};
+    private touched = false;
+    private disabled = false;
+
+    set quantity(value: number) {
+        this.markAsTouched();
+        if (this.disabled) { return; }
+
+        this._quantity = this.sanitizeValue(value);
+        this.onChange(this._quantity);
+    }
+
+    get quantity(): number {
+        return this._quantity;
+    }
+
+    constructor() {}
+
+    writeValue(value: number): void {
+        this.quantity = value;
+    }
+
+    registerOnChange(fn: any): void {
+        this.onChange = fn;
+    }
+
+    registerOnTouched(fn: any): void {
+        this.onTouched = fn;
+    }
+
+    setDisabledState(disabled: boolean): void {
+        this.disabled = disabled;
+    }
+
+    onIncrement(): void {
+        this.quantity = this.quantity + 1;
+    }
+
+    onDecrement(): void {
+        this.quantity = this.quantity - 1;
+    }
+
+    onUserChange(event: Event): void {
+        const value = (event.target as HTMLInputElement).valueAsNumber;
+        this.quantity = value;
+    }
+
+    validate(control: AbstractControl): ValidationErrors | null {
+        const value = control.value;
+
+        if (this.minValue != null && value < this.minValue) {
+            return { mustBeAtLeastMin: value };
+        }
+
+        if (this.maxValue != null && value > this.maxValue) {
+            return { mustBeLessThanMax: value };
+        }
+
+        return null;
+    }
+
+    private markAsTouched(): void {
+        if (this.touched) { return; }
+
+        this.onTouched();
+        this.touched = true;
+    }
+
+    private sanitizeValue(value: number): number {
+        return isNaN(value)
+            ? this.defaultValue
+            : this.clampToRange(value);
+    }
+
+    private clampToRange(value: number): number {
```

Comment:
I dont believe these two methods are required:

1) sanitizing is not required because we define a custom number value and all types only take numbers as inputs so it is redundant to check for isNan
2) clampToRange already exists as a pipe. We should instead import the pipe in this component and call .transform in the setter

Reply:
The reason I check for NaN is because the user can type their own value into the input element in the HTML. If they type characters, nothing will be entered into the box, but it will return a blank value to the code here and cause the browser to spit out an error about trying to parse a NaN value, leaving the input empty (which, if they then clicked "Add to cart" I'm fairly sure would cause an error with Snipcart as well).

Reply:
the type attribute does that for us. Also, it would make more sense to disable typing characters altogether by disabling the input